svelte-web-frontend/src/lib/ai/compute-motion-descriptors.ts

```diff
@@ -0,0 +1,190 @@
+ */
+export function removeDuplicateFrameTimes(userPoses: Pose2DPixelLandmarks[], frameTimes: number[]): { adjUserPoses: Pose2DPixelLandmarks[], uniqueFrameTimes: number[] } {
+    const uniqueFrameTimes = [];
+    const adjUserPoses = [];
+
+    for (let i = 0; i < frameTimes.length; i++) {
+        const currentTime = frameTimes[i];
+
+        // Check if the current frame time is not the same as the previous one
+        if (i === 0 || currentTime !== frameTimes[i - 1]) {
+            uniqueFrameTimes.push(currentTime);
+            adjUserPoses.push(userPoses[i]);
+        }
+    }
+
+    return { adjUserPoses, uniqueFrameTimes };
+}
+
+
+/**
+ * Calculates the standard deviation of the differences between adjusted user jerks and reference jerks.
+ *
+ * @param {Pose2DPixelLandmarks[]} adjUserPoses - Array of adjusted user poses.
+ * @param {Pose2DPixelLandmarks[]} referencePoses - Array of reference poses.
+ * @param {number[]} uniqueFrameTimes - Array of unique frame times corresponding to the poses.
+ * @returns {[number, number, number, number, number, number]} MSE and RMSE between user and reference motion descriptors.
+ *
+ * @throws {Error} Throws an error if there is invalid input data, mismatched array lengths, or bad unique frametimes.
+ */
+export function calculateMotionDescriptorsScore(
+  adjUserPoses: Pose2DPixelLandmarks[],
+  referencePoses: Pose2DPixelLandmarks[],
+  uniqueFrameTimes: number[]
+): [number, number, number, number, number, number] {
+  if (!adjUserPoses || adjUserPoses.length === 0 || !referencePoses || referencePoses.length === 0) {
+      throw new Error("Invalid input data. Both adjUserPoses and referencePoses must be non-empty arrays.");
+  }
+
+  if (adjUserPoses.length !== referencePoses.length) {
+      throw new Error("Mismatched array lengths between adjUserPoses and referencePoses.");
+  }
+
+  const usrScale = GetScaleIndicator(adjUserPoses[0]);
+  const refScale = GetScaleIndicator(referencePoses[0]);
+
+  if (!uniqueFrameTimes || uniqueFrameTimes.length === 0) {
+      throw new Error("Bad uniqueFrameTimes");
+  }
+
+  const userVels = calculateVels(adjUserPoses, uniqueFrameTimes, usrScale);
+  const referenceVels = calculateVels(referencePoses, uniqueFrameTimes, refScale);
+  const userAccs = calculateDerivative(userVels, uniqueFrameTimes);
+  const referenceAccs = calculateDerivative(referenceVels, uniqueFrameTimes);
+  const userJerks = calculateDerivative(userAccs, uniqueFrameTimes);
+  const referenceJerks = calculateDerivative(referenceAccs, uniqueFrameTimes);
+
+  const usrScalarVels = calculateMetricsMag(userVels);
+  const refScalarVels = calculateMetricsMag(referenceVels);
+  const usrScalarAccs = calculateMetricsMag(userAccs);
+  const refScalarAccs = calculateMetricsMag(referenceAccs);
+  const usrScalarJerks = calculateMetricsMag(userJerks);
+  const refScalarJerks = calculateMetricsMag(referenceJerks);
+
+  const jerksMAE = getMatricesMAE(usrScalarJerks, refScalarJerks);
```

Comment:
Can you add a comment explaining why we're calculating the MAE and RMSE metrics, and a bit about how we would interpret each values and what that output range might be?

Reply:
Done.